### PR TITLE
Add more cases to BenchmarkPartitionedOutputOperator

### PR DESCRIPTION
**Pull request by lukasz-stec**

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 1]
```diff
@@ -285,9 +286,9 @@ private PartitionedOutputBuffer createPartitionedOutputBuffer()
 
         private PartitionedOutputOperator createPartitionedOutputOperator()
         {
-            PartitionFunction partitionFunction = new LocalPartitionGenerator(
-                    new PrecomputedHashGenerator(0),
-                    partitionCount);
+            PartitionFunction partitionFunction = new BucketPartitionFunction(
+                    HASH.createBucketFunction(null, true, partitionCount, null),
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:385]: extract `HashBucketFunction` instead as a public class

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 2]
```diff
@@ -549,8 +549,13 @@ public static Block createLongSequenceBlock(int start, int end)
     public static Block createLongDictionaryBlock(int start, int length)
     {
         checkArgument(length > 5, "block must have more than 5 entries");
+        return createLongDictionaryBlock(start, length, length / 5);
+    }
+
+    public static Block createLongDictionaryBlock(int start, int length, int dictionarySize)
+    {
+        checkArgument(dictionarySize > 0, "dictionarySize must be more than 0");
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:557]: `must be more` -> `must be greater`

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 3]
```diff
@@ -549,8 +549,13 @@ public static Block createLongSequenceBlock(int start, int end)
     public static Block createLongDictionaryBlock(int start, int length)
     {
         checkArgument(length > 5, "block must have more than 5 entries");
+        return createLongDictionaryBlock(start, length, length / 5);
+    }
+
+    public static Block createLongDictionaryBlock(int start, int length, int dictionarySize)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:555]: I think you should change meaning of arguments. Make `length` the actual length of underlying dictionary (that would be consistent with other methods) and make `dictionarySize` length of top-level block

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 4]
```diff
@@ -549,8 +549,13 @@ public static Block createLongSequenceBlock(int start, int end)
     public static Block createLongDictionaryBlock(int start, int length)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:549]: Could you add numbers as part of this commit compared to non-dictionary blocks?

Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:559]: Do you mean benchmark  `DICTIONARY_HASH_BIGINT_20_PERCENT` against the same blocks but with hash block being direct `LongArrayBlock` instead `DictionaryBlock`?

**Review by lukasz-stec — COMMENTED** [review]
New case with skewed hash block added

File: core/trino-main/src/test/java/io/trino/operator/PageTestUtils.java [hunk 5]
```diff
@@ -83,11 +83,15 @@ public static Page createRandomPage(
             blocks.add(wrapping.map(w -> w.wrap(block, positionCount)).orElse(block));
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/PageTestUtils.java:83]: add to commit message what the bug was about

File: core/trino-main/src/main/java/io/trino/sql/planner/SystemPartitioningHandle.java [hunk 6]
```diff
@@ -261,7 +261,7 @@ public String toString()
             }
         }
 
-        private static class HashBucketFunction
+        public static class HashBucketFunction
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/SystemPartitioningHandle.java:264]: Also make it top level class so it's not hidden between private classes and other stuff

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 7]
```diff
@@ -25,14 +25,14 @@
 import io.trino.jmh.Benchmarks;
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:25]: Add to commit message why it the right generation compared to `LocalPartitionGenerator`

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java (hunk 4 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:559]: Yeah. Against same blocks but without dictionaries. But it's a nit, you can skip it

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 8]
```diff
@@ -227,6 +230,13 @@ public static Block createRandomLongDecimalsBlock(int positionCount, float nullR
                 () -> String.valueOf(RANDOM.nextLong())));
     }
 
+    public static Block createRandomLongsBlock(int positionCount, int numberOfUniqueValues)
+    {
+        checkArgument(positionCount >= numberOfUniqueValues, "numberOfUniqueValues must be between 1 and positionCount: %s but was %s", positionCount, numberOfUniqueValues);
+        int[] randomPartitions = chooseRandomUnique(positionCount, numberOfUniqueValues).toIntArray();
+        return createLongsBlock(IntStream.range(0, positionCount).mapToLong(position -> randomPartitions[RANDOM.nextInt(numberOfUniqueValues)]).boxed().collect(toImmutableList()));
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:237]: 1. Why not just pick random value from `[0, numberOfUniqueValues)` it's as good as this solution
nit: this doesn't really guarantee that there will be really `numberOfUniqueValues` unique values.

Anyway, from processing POV long block with `numberOfUniqueValues` and all random values shouldn't be that different (apart from memory locality in joins maybe). Maybe just remove this method?

2. move `.mapToLong` `.boxed` to separate lines

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java (hunk 3 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:565]: Ok Let's keep it as is

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 9]
```diff
@@ -175,18 +193,96 @@ public void verifyAddPage()
...
+
             DICTIONARY_BIGINT(BigintType.BIGINT, 3000) {
                 @Override
                 public Page createPage(List<Type> types, int positionCount, float nullRate)
                 {
-                    return createRandomDictionaryPage(types, POSITION_COUNT, nullRate);
+                    return createRandomDictionaryPage(types, positionCount, nullRate);
                 }
             },
+
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:214]: undo

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 10]
```diff
@@ -175,18 +193,96 @@ public void verifyAddPage()
...
             RLE_BIGINT(BigintType.BIGINT, 3000) {
                 @Override
                 public Page createPage(List<Type> types, int positionCount, float nullRate)
                 {
-                    return createRandomRlePage(types, POSITION_COUNT, nullRate);
+                    return createRandomRlePage(types, positionCount, nullRate);
+                }
+            },
+            DICTIONARY_HASH_BIGINT_20_PERCENT(BigintType.BIGINT, 3000) {
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:222]: what does it mean `dictionary_hash`. Neither of the columns are hash of the other one

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java (hunk 8 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java]: If we only pick random from  `[0, numberOfUniqueValues)` as a value and then partition by that block, we will have partitions with non zero elements only at the first `numberOfUniqueValues` positions
e.g. `[4,5,3,0,0,0,0,0,0,0]` where in the real world I expect something like `[0,0,4,5,0,0,03,0,0]`.
The first pattern is easy to predict by for example branch predictor, which may skew benchmarking results.

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java (hunk 10 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:220]: I'm using hash channels and partition channels interchangeably. Does `DICTIONARY_PARTITION_...` sound better?

Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:220]: > Does DICTIONARY_PARTITION_... sound better?

Yup. Although I don't know what is the difference between this and `DICTIONARY_BIGINT`

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 11]
```diff
@@ -771,11 +789,19 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
         return chooseRandomUnique(positionCount, nullCount);
     }
 
-    private static Set<Integer> chooseRandomUnique(int bound, int count)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:774]: why this change. Please undo

File: core/trino-main/src/main/java/io/trino/sql/planner/HashBucketFunction.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/HashBucketFunction.java:1]: `Make HashBucketFunction public and extract` -> `Make HashBucketFunction public and extract as top level class`

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 13]
```diff
@@ -227,6 +229,16 @@ public static Block createRandomLongDecimalsBlock(int positionCount, float nullR
                 () -> String.valueOf(RANDOM.nextLong())));
     }
 
+    public static Block createRandomLongsBlock(int positionCount, int numberOfUniqueValues)
+    {
+        checkArgument(positionCount >= numberOfUniqueValues, "numberOfUniqueValues must be between 1 and positionCount: %s but was %s", positionCount, numberOfUniqueValues);
+        int[] randomPartitions = chooseRandomUnique(positionCount, numberOfUniqueValues).stream().mapToInt(Integer::intValue).toArray();
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:235]: 1. move `.mapToInt` and `.toArray()` to separate lines
2. Rename `randomPartitions` to `uniqueValues`

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java (hunk 11 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:774]: > I would still leave the fast path though as my heart breaks when 8k numbers have to be generated and shuffled just to take the first two...

IMO it's not really important since it's not a perf critical code so simplicity > optimizations

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 14]
```diff
@@ -773,6 +785,14 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
 
     private static Set<Integer> chooseRandomUnique(int bound, int count)
     {
+        if (count < bound / 10) {
+            Set<Integer> values = new HashSet<>(count);
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:794]: If you want to keep it, add a short comment why it's there

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 15]
```diff
@@ -32,6 +32,7 @@
 import java.util.ArrayList;
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:32]: Make commit message more explicit, e.g: `Add createRandomLongsBlock using values from unique values set`

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 16]
```diff
@@ -145,10 +154,21 @@ public void verifyAddPage()
...
                 "BIGINT",
+                "BIGINT_SKEWED_HASH",
                 "DICTIONARY_BIGINT",
                 "RLE_BIGINT",
+                "BIGINT_PARTITION_BIGINT_20_PERCENT",
+                "DICTIONARY_PARTITION_BIGINT_20_PERCENT",
+                "DICTIONARY_PARTITION_BIGINT_50_PERCENT",
+                "DICTIONARY_PARTITION_BIGINT_80_PERCENT",
+                "DICTIONARY_PARTITION_BIGINT_100_PERCENT",
+                "RLE_HASH_BIGINT",
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:170]: `RLE_HASH_BIGINT` -> `RLE_PARTITION_BIGINT`?

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 17]
```diff
@@ -264,15 +394,28 @@ public void setup(Blackhole blackhole)
...
         }
 
+        private static Page page(int positionCount, int channelCount, Supplier<Block> standardBlock, Block partitionBlock)
+        {
+            ImmutableList.Builder<Block> blocks = ImmutableList.builder();
+            for (int i = 0; i < channelCount; i++) {
+                blocks.add(standardBlock.get());
+            }
+            blocks.add(partitionBlock);
+            return new Page(
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:414]: Hash should be precalculated too. We rather don't want to benchmark interpreted hash generator since it will shadow perf of partition output operator itself

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 18]
```diff
@@ -175,18 +197,114 @@ public void verifyAddPage()
...
+            },
+            BIGINT_PARTITION_BIGINT_20_PERCENT(BigintType.BIGINT, 3000) {
+                @Override
+                public Page createPage(List<Type> types, int positionCount, float nullRate)
+                {
+                    return page(
+                            positionCount,
+                            types.size(),
+                            () -> createRandomBlockForType(BigintType.BIGINT, positionCount, nullRate),
+                            createLongsBlock(LongStream.range(0, positionCount)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:233]: why don't you use generator method that uses unique values

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 19]
```diff
@@ -175,18 +197,114 @@ public void verifyAddPage()
...
+                                    .mapToObj(value -> value % (positionCount / 5))
+                                    .collect(toImmutableList())));
+                }
+            },
+            DICTIONARY_PARTITION_BIGINT_20_PERCENT(BigintType.BIGINT, 3000) {
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:238]: `...` -> `BIGINT_DICTIONARY_PARTITION_CHANNEL_20_PERCENT` I think it's more explicit (english) this way

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java (hunk 18 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:233]: I need only `positionCount/5 `unique values but `postitionCount` total values

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java (hunk 17 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:407]: I may be missing something but here we only create a page with given `partitionBlock` as the last block, which is gonna be configured in the operator as (precomputed) partition/hash channel.

File: core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java [hunk 20]
```diff
@@ -213,17 +331,22 @@ public Page createPage(List<Type> types, int positionCount, float nullRate)
 
             public Page createPage(List<Type> types, int positionCount, float nullRate)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/BenchmarkPartitionedOutputOperator.java:332]: nit: `types` argument is not really needed, you could pass just channel count

**Review by sopel39 — APPROVED** [review]
